File: src/main/java/singleton_test/LazyThreadSafeDoubleCheck.java
package singleton_test;

/**
 * Created by Alex on 2017/5/27.
 * 4.懒加载，双重同步锁
 * Alex
 */

public class LazyThreadSafeDoubleCheck {

    private static volatile LazyThreadSafeDoubleCheck instance ;

    private LazyThreadSafeDoubleCheck(){
        System.out.println("懒加载，双重同步锁");
    }

    public static LazyThreadSafeDoubleCheck getInstance(){
        if (instance == null){
            synchronized (LazyThreadSafeDoubleCheck.class){
                if (instance == null){
                    instance = new LazyThreadSafeDoubleCheck();
                }
            }
        }
        return instance;
    }

}

File: src/main/kotlin/kotlin_singleton_test/LazyThreadSafeDoubleCheck.kt
package kotlin_singleton_test

/**
 * Created by Alex on 2017/5/27.
 * 4.懒加载，双重同步锁
 * Alex
 */
class LazyThreadSafeDoubleCheck private constructor(val msg:String){
    init {
        println(msg)
    }
    companion object{
        //kotlin常规
        val instance by lazy(LazyThreadSafetyMode.SYNCHRONIZED){
            LazyThreadSafeDoubleCheck("kotlin常规：懒加载，双重同步锁")
        }
        //模仿JAVA
        @Volatile private var instance2: LazyThreadSafeDoubleCheck? = null

        fun get(): LazyThreadSafeDoubleCheck {
            if (instance2 == null) {
                synchronized(this) {
                    if (instance2 == null) {
                        instance2 = LazyThreadSafeDoubleCheck("kotlin模仿JAVA：懒加载，双重同步锁")
                    }
                }
            }
            return instance2!!
        }
    }
}